cmd/rehearse/main.go

```diff
@@ -0,0 +1,331 @@
+
+	return &rehearsal, nil
+}
+
+func loadClusterConfig() (*rest.Config, error) {
+	clusterConfig, err := rest.InClusterConfig()
+	if err == nil {
+		return clusterConfig, nil
+	}
+
+	credentials, err := clientcmd.NewDefaultClientConfigLoadingRules().Load()
+	if err != nil {
+		return nil, fmt.Errorf("could not load credentials from config: %v", err)
+	}
+
+	clusterConfig, err = clientcmd.NewDefaultClientConfig(*credentials, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not load client configuration: %v", err)
+	}
+	return clusterConfig, nil
+}
+
+func submitRehearsal(job *config.Presubmit, jobSpec *pjapi.ProwJobSpec, logger logrus.FieldLogger, pjclient pj.ProwJobInterface, dry bool) (*pjapi.ProwJob, error) {
+	labels := make(map[string]string)
+	for k, v := range job.Labels {
+		labels[k] = v
+	}
+
+	pj := pjutil.NewProwJob(pjutil.PresubmitSpec(*job, *(jobSpec.Refs)), labels)
+	logger.WithFields(pjutil.ProwJobFields(&pj)).Info("Submitting a new prowjob.")
+
+	if dry {
+		jobAsYAML, err := yaml.Marshal(pj)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to marshal job to YAML: %v", err)
+		}
+		fmt.Printf("%s\n", jobAsYAML)
+		return &pj, nil
+	}
+
+	return pjclient.Create(&pj)
+}
+
+// Rehearsed Prow jobs may depend on ConfigMaps with content also modified by
+// the tested PR. All ci-operator-based jobs use the `ci-operator-configs`
+// ConfigMap that contains ci-operator configuration files. Rehearsed jobs
+// need to have the PR-version of these files available. The following code
+// takes care of creating a short-lived, rehearsal ConfigMap. The keys needed
+// to be present are extracted from the rehearsal jobs and the rehearsal jobs
+// are modified to use this ConfigMap instead of the "production" one.
+
+var ciOperatorConfigsCMName = "ci-operator-configs"
+
+type rehearsalCIOperatorConfigs struct {
+	cmclient  corev1.ConfigMapInterface
+	prNumber  int
+	configDir string
+
+	logger logrus.FieldLogger
+	dry    bool
+
+	configMapName string
+	neededConfigs map[string]string
+}
+
+func newRehearsalCIOperatorConfigs(cmclient corev1.ConfigMapInterface, prNumber int, configDir string, logger logrus.FieldLogger, dry bool) *rehearsalCIOperatorConfigs {
+	name := fmt.Sprintf("rehearsal-ci-operator-configs-%d", prNumber)
+	return &rehearsalCIOperatorConfigs{
+		cmclient:      cmclient,
+		prNumber:      prNumber,
+		configDir:     configDir,
+		logger:        logger.WithField("ciop-configs-cm", name),
+		dry:           dry,
+		configMapName: name,
+		neededConfigs: map[string]string{},
+	}
+}
+
+// If a job uses the `ci-operator-config` ConfigMap, save which key does it use
+// from it and replace that ConfigMap reference with a reference to the
+// temporary, rehearsal ConfigMap containing the necessary keys with content
+// matching the version from tested PR
+func (c *rehearsalCIOperatorConfigs) FixupJob(job *config.Presubmit, repo string) {
```

Comment:
But all the keys will be mounted in the pod that `pj-rehearser` will run, right?

Reply:
Not sure I understand the question. This temporary CM only contains the subset of keys that the jobs being rehearsed need, and each individual job only mounts the keys it needs. The rehearsal jobs mount the same keys as "production" jobs, just not from the original `ci-operator-config` CM, but from the rehearsal CM.

Reply:
I plan to cover this code by unit tests that might provide some helpful examples of how the jobs are fixed up and how the resulting CM looks like.